File: examples/lib/simple_workflow.rs
//! Basic workflow execution example
//!
//! Run with: cargo run --example simple_workflow

use jackdaw::DurableEngineBuilder;
use serverless_workflow_core::models::{
    task::{SetTaskDefinition, SetValue, TaskDefinition},
    workflow::{DocumentMetadata, WorkflowDefinition},
};
use std::time::Duration;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let engine = DurableEngineBuilder::new().build()?;

    let workflow = WorkflowDefinition {
        document: DocumentMetadata {
            dsl: "1.0.2".to_string(),
            namespace: "examples".to_string(),
            name: "simple-transform".to_string(),
            version: "1.0.0".to_string(),
            ..Default::default()
        },
        do_: vec![
            (
                "greet".to_string(),
                TaskDefinition::Set(SetTaskDefinition {
                    set: SetValue::Inline(serde_json::json!({
                        "greeting": "Hello, World!",
                        "timestamp": "${ now() }"
                    })),
                    ..Default::default()
                }),
            ),
            (
                "transform".to_string(),
                TaskDefinition::Set(SetTaskDefinition {
                    set: SetValue::Inline(serde_json::json!({
                        "message": "${ \"Greeting: \" + .greeting }",
                        "time": "${ .timestamp }"
                    })),
                    ..Default::default()
                }),
            ),
        ]
        .into_iter()
        .collect(),
        ..Default::default()
    };

    let input = serde_json::json!({ "user": "Alice" });
    let handle = engine.execute(workflow, input).await?;
    let result = handle.wait_for_completion(Duration::from_secs(30)).await?;

    println!("{}", serde_json::to_string_pretty(&result)?);

    Ok(())
}

File: examples/lib/python/simple_workflow.py
#!/usr/bin/env python3
"""
Simple workflow execution example using jackdaw Python bindings.

Build the Python module first:
    maturin develop --features python

Run this example:
    python examples/python/simple_workflow.py
"""

import asyncio
import jackdaw

WORKFLOW_YAML = """
document:
  dsl: '1.0.2'
  namespace: examples
  name: simple-transform
  version: 1.0.0
do:
  - greet:
      set:
        greeting: "Hello, World!"
        user: "${ .user }"
  - transform:
      set:
        message: "${ \\"Greeting: \\" + .greeting }"
        full_message: "${ .message + \\" User: \\" + .user }"
"""

async def main():
    builder = jackdaw.DurableEngineBuilder()
    engine = builder.build()

    input_data = {"user": "Alice"}

    handle = await engine.execute(WORKFLOW_YAML, input_data)

    print(f"Workflow instance ID: {handle.instance_id()}")

    result = await handle.wait_for_completion(30.0)

    print(f"Result: {result}")

if __name__ == "__main__":
    asyncio.run(main())
